### ISSUE TITLE
Performance on imbalanced dataset

### ISSUE THREAD

**Issue by zekidsp** [tech-srl/code2vec#89]
Hi 

We are using your code2vec method for a multilabel binary classification task on a highly imbalanced dataset containing a lot of  function-level codes.  The functions are labeled according 5 binary categories.  The ratio between positive and negative samples is around 5:95 or higher for all categories, which makes this task very challenging.  We tried alternative experiment settings such as (i) extracting codevectors while keeping method/function names as in the original case and using  the extracted codevectors in another ML algorithm for multilabel classification, (ii) replacing method names with corresponding collective  labels of 5 categories  and then predicting new method names by using the code2vec, and (iii)  replacing  method names with the corresponding single label for each category and performing prediction separately for each category.  Unfortunately, the results we obtained in all cases are not very good. Could this be due to high imbalance in the dataset? Any suggestion to improve prediction performance? Should we change any parameter in code2vec setting for this high imbalance situation?   

Thank you in advance.

**Comment by urialon** [comment]
Hi @zekidsp,
Thank you for your interest in code2vec!

You should probably change the metrics that code2vec measures, to check
where do the model's mistake stem from. Does the model suffer from
false-negative or false-positives?

There are some common techniques for handling imbalanced data. For example,
have you tried weighting the positive examples according to it's inverse
ratio?

Additionally, you can try our code2SEQ model which has a much better
LSTM-based  encoder, even when your output is a single class (a sequence of
length=1).

Best,
Uri

**Comment by zekidsp** [comment]
Thank you @urialon for your prompt response.

Yes, we get high false-negative and false-positives.  Is there a way to set weights for positive samples in the code2vec? 
I will look at the code2SEQ. 

BR
Zeki

**Comment by Makifer** [comment]
Hi @urialon , 
On a similar topic I have a question: is the input format for code2vec and code2seq projects are the same? Can we use the same preprocessed inputs on both of them without any change? I could not see any difference myself from the project descriptions.

**Comment by urialon** [comment]
Hi,
There is no implemented way of setting weights, but you can try to
decompose the expression here:
https://github.com/tech-srl/code2vec/blob/master/tensorflow_model.py#L228

into multiple steps:
first, compute the `tf.nn.sparse_softmax_cross_entropy_with_logits`. Then
multiply the resulting scores by `(tf.one_hot(input_tensors.target_index) *
19) + 1` such that the positive examples will be multiplied by 20, and the
negative examples will be multiplied by 1. Only afterward, apply the `tf.
reduce_sum`.

Best,
Uri

**Comment by urialon** [comment]
Hi @Makifer ,
I think that the input for code2**seq** can be used for code2**vec**, but not vice-versa.

The reason is that in code2**seq** the path is split to nodes, e.g., `ForStatement|IfExpression|VarDeclaration ...` while in code2**vec** the entire path is hashed into a single string of digits, like `1234567`.

Additionally, the statistics and histograms of the different models are computed slightly differently, and also stored slightly differently. 

So in general, if you created a new custom dataset you can use it for both code2vec and code2seq, but just follow the preprocessing scripts carefully to verify that the histograms and dictionaries are computed and stored the same.

**Comment by Makifer** [comment]
Hi @urialon , 
Thanks for explanation.

In our experiment, we created a new C++ dataset using https://github.com/JetBrains-Research/astminer library which generate paths. **astminer** is referenced both in [code2seq](https://github.com/tech-srl/code2seq/#extending-to-other-languages) and in [code2vec](https://github.com/tech-srl/code2vec/#extending-to-other-languages). So, we think that we can use the dataset with code2seq as well.

However astminer generates paths like this: `12,885,72` where 12 is the first node_id, 885 is the path_id, and 72 is the last node_id. When we look at the https://github.com/tech-srl/code2seq/blob/master/README.md#java , we noticed that, in the preprocessed datasets, paths include token names instead of token_ids so we replace them in our generated dataset as well and we used those paths as input to code2vec.
To be clear, our paths look like this: `int,885,fprintf` where 885 is the path_id. Should we also put the entire path description to be able to use it with code2seq? Or the current version should be fine?

**Comment by urialon** [comment]
Using token ids instead of token names is fine. So you don't even need to
replace them in code2vec.

In code2**seq** however, the network expects the tokens to be split to
subtokens, and the paths to be split to nodes.
So for example, code2**seq** expects an input like:
`my|str,If|For|Expression|Statement,new|string|value` where `my|str` is the
token value, `If|For|Expression|Statement` is the path, and
`new|string|value` is the token value in the other side of the path. So
this format **is supposed to work** in code2**seq** as well, but will not
leverage the power of code2seq over code2vec.

Does that make sense?

**Comment by urialon** [comment]
Closing due to inactivity, feel free to re-open if you have further questions.
Uri
